### PR TITLE
Fix bug 18941: Marshal.OffsetOf and Marshal.SizeOf return incorrect results

### PR DESCRIPTION
**Pull request by kg**
This fixes https://bugzilla.xamarin.com/show_bug.cgi?id=18941. Our existing implementation of Marshal.OffsetOf and Marshal.SizeOf is incorrect for empty structs - the windows .NET framework pads them up to 1 byte in size, while we do not. This also impacts the marshaling layout of larger structs (if they contain 0-byte structs).

Our in-memory layout of these structs also differs but @vargaz said that shouldn't be an issue, so I opted to leave that out of the automated test and not fix it. (Fixing it seems prohibitively difficult anyway.) The impact of that existing bug is that some of the other Marshal methods (like the one for computing the offset of individual array elements) will return results that don't match windows .NET framework. The impact of that is much smaller in practice because I believe it would require broken code that pins managed arrays and expects element sizes to match their marshal.sizeof (instead of using pointer arithmetic).

### CODE REVIEW

File: mono/metadata/marshal.c [hunk 1]
```diff
@@ -11180,7 +11181,8 @@ ves_icall_System_Runtime_InteropServices_Marshal_SizeOf (MonoReflectionTypeHandl
 		return 0;
 	}
 
-	return mono_class_native_size (klass, NULL);
+	result = mono_marshal_type_size (type, NULL, &align, FALSE, klass->unicode);
+	return (guint32)result;
```
Comment (vargaz) [inline mono/metadata/marshal.c]: Why is this needed ?

Comment (kg) [reply mono/metadata/marshal.c]: The native size is 0 for an empty struct and for a struct containing empty structs the size will also be transitively wrong

Comment (vargaz) [reply mono/metadata/marshal.c]: Shouldn't the native size be 1 as well ?

Comment (kg) [reply mono/metadata/marshal.c]: It's not, it's 0 everywhere in the runtime. It's 1 on .NET framework.

Comment (jaykrell) [reply mono/metadata/marshal.c]: The behavior is visible, and compatibility with mono and .net is impossible? Or behavior is only visible when marshaling? Persistant data compat?

Comment (luhenry) [reply mono/metadata/marshal.c]: Do we have a test covering this behavior? If so I would be ok with merging it.

Comment (kg) [reply mono/metadata/marshal.c]: The test I added covers this (the test won't pass without it) but I don't know how good our coverage is otherwise.

Comment (vargaz) [reply mono/metadata/marshal.c]: I mean mono_class_native_size () should return the same value as mono_marshal_type_size (), mono_marshal_type_size () calls mono_class_native_size () for structs:

	case MONO_NATIVE_STRUCT: 
		klass = mono_class_from_mono_type (type);
		if (klass == mono_defaults.object_class &&
			(mspec && mspec->native == MONO_NATIVE_STRUCT)) {
		*align = 16;
		return 16;
		}
		return mono_class_native_size (klass, align);

Comment (luhenry) [reply mono/metadata/marshal.c]: But `mono_marshal_type_size` has been updated at https://github.com/mono/mono/pull/7142/files#diff-559b8405966c917ff45da9a2a9bf0296R11942 to guarantee a size of at least 1.

**Comment by luhenry** [comment]
Fixes https://github.com/mono/mono/projects/3#card-7346260

File: mono/metadata/marshal.c (hunk 1 above)
Comment (vargaz) [reply mono/metadata/marshal.c]: Wouldn't it be better to do this in mono_class_native_size () ?

Comment (luhenry) [reply mono/metadata/marshal.c]: I don't know, maybe? It seems like `mono_class_native_size` is used in many more places than `mono_marshal_type_size` so I don't know what's the impact of such a change. How about we merge here and figure out what else we can transition from `mono_class_native_size` to `mono_marshal_type_size` in a follow-up PR, and possibly merge both.

Comment (kg) [reply mono/metadata/marshal.c]: My original attempted fix was to make the native size of empty structs 1 byte, as in .net framework. It broke a bunch of things, in part because it seems like we have multiple copies of size/offset calculation logic in different parts of mono and I failed to find and fix all of them.

**Comment by marek-safar** [comment]
@vargaz are you ok with the change?
